### PR TITLE
BYOM for local

### PR DESCRIPTION
**Pull request by ea-rus**
## Description

BYOM enabled locally.

### Http api:

**POST** api/handlers/byom/<engine_name>

Uploaded files
- code - python file with model code
- modules - text file with requirements

**Logic:**

1. Create new ml engine
2. If virtualenv is installed on computer:
   - try to create a new virtual environment using current python interpreter
   - install all requirements for model
     - if pandas is not in requirements - it is installed anyway 
   - install pyarrow (for sending dataframe between different pandas versions)
3. If not virtualenv is installed:
   - try to install requirements to current environment
4. Try to import uploaded model and check predict and train methods
5. If not success:
   - remove virtual environment if it was created
   - remove ml engine

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📄 This change requires a documentation update

### What is the solution?

(Describe at a high level how the feature was implemented)

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation, or created issues to update them.
- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] I have shared a short loom video or screenshots demonstrating any new functionality.

### CODE REVIEW

File: mindsdb/integrations/handlers/byom_handler/byom_handler.py [hunk 1]
```diff
@@ -1,28 +1,174 @@
...
+    def predict(self, df, args=None):
+
+        model_proxy = self._get_model_proxy()
+
+        model_state = self.model_storage.file_get('model')
+
+        pred_df = model_proxy.predict(df, model_state)
+
+        # rename target column
+        # target = self.model_storage.get_info()['to_predict'][0]
```
Comment (paxcema) [inline mindsdb/integrations/handlers/byom_handler/byom_handler.py:73]: Leftover comments?

File: mindsdb/integrations/handlers/byom_handler/byom_handler.py [hunk 2]
```diff
@@ -1,28 +1,174 @@
...
+        # for dataframe serialization
+        modules.append('pyarrow==11.0.0')
+        return modules
+
+    def install_modules(self, modules):
+        # install in current environment using pip
+
+        pip_cmd = self.python_path.parent / 'pip'
+        for module in modules:
+            os.system(f'{pip_cmd} install {module}')
```
Comment (paxcema) [inline mindsdb/integrations/handlers/byom_handler/byom_handler.py:164]: Maybe add an error check to raise an exception if something goes wrong (e.g. the user specified some package version that is not available for his specific OS version).

**Review by paxcema — APPROVED** [review]

File: mindsdb/integrations/handlers/byom_handler/byom_handler.py (hunk 1 above)
Comment (ea-rus) [reply mindsdb/integrations/handlers/byom_handler/byom_handler.py]: removed

File: mindsdb/integrations/handlers/byom_handler/byom_handler.py (hunk 2 above)
Comment (ea-rus) [reply mindsdb/integrations/handlers/byom_handler/byom_handler.py:161]: what do you think about replacing it to?:
`if os.system('...') != 0: raise`

Comment (paxcema) [reply mindsdb/integrations/handlers/byom_handler/byom_handler.py:161]: Yes, that should work I think 👍

**Comment by paxcema** [comment]
@ea-rus would this template work for time series BYOMs?

**Comment by ea-rus** [comment]
@paxcema, it should work with time series as well. Predict method receives one dataframe with all groups and has to split it to make forecast for every group.

**Pull request merged**